osu.Game/Beatmaps/Beatmap.cs

```diff
@@ -48,6 +48,44 @@ public class Beatmap<T> : IBeatmap<T>
 
         public virtual IEnumerable<BeatmapStatistic> GetStatistics() => Enumerable.Empty<BeatmapStatistic>();
 
+        public double GetMostCommonBeatLength()
+        {
+            // The last playable time in the beatmap - the last timing point extends to this time.
+            // Note: This is more accurate and may present different results because osu-stable didn't have the ability to calculate slider durations in this context.
+            double lastTime = HitObjects.LastOrDefault()?.GetEndTime() ?? ControlPointInfo.TimingPoints.LastOrDefault()?.Time ?? 0;
+
+            var beatLengthsAndDurations =
+                // Construct a set of (beatLength, duration) tuples for each individual timing point.
+                ControlPointInfo.TimingPoints.Select((t, i) =>
+                                {
+                                    if (t.Time > lastTime)
+                                        return (beatLength: t.BeatLength, 0);
+
+                                    var nextTime = i == ControlPointInfo.TimingPoints.Count - 1 ? lastTime : ControlPointInfo.TimingPoints[i + 1].Time;
+                                    return (beatLength: t.BeatLength, duration: nextTime - t.Time);
+                                })
+                                // Aggregate durations into a set of (beatLength, duration) tuples for each beat length
+                                .GroupBy(t => t.beatLength)
+                                .Select(g => (beatLength: g.Key, duration: g.Sum(t => t.duration)))
+                                // And if there are no timing points, use a default.
+                                .DefaultIfEmpty((TimingControlPoint.DEFAULT_BEAT_LENGTH, 0));
```

Comment:
Is there a reason you can't end this with `.OrderByDescending(i => i.Duration).FirstOrDefault()` rather than the code following this?

Reply:
I was expecting a comment in the opposite direction, feeling uneasy about the extra cost of sorting vs a linear lookup. If you're fine with that I have no problem.

Reply:
i don't think it's much more expensive at the point you're iterating the whole enum. probably not worth worrying about until we realise that the linq overhead (at all) is an issue, and rewrite without any.

Reply:
I'm not sure whether there are crazy maps out there with hundreds of timing points (there probably are, as always), but I'd like to think even linq sorting shouldn't be worse than n log n? I don't think that will be super problematic either, but I may well be wrong.

Reply:
I'd still propose this change. For rare executions like this there should be seldom a case where linq becomes an issue (and at the point it does, we should probably be converting the whole code to non-linq).